src/render/SDL_render.c

```diff
@@ -5008,11 +4954,13 @@ bool SDL_RenderPresent(SDL_Renderer *renderer)
 
     CHECK_RENDERER_MAGIC(renderer, false);
 
-    if (renderer->logical_target) {
-        SDL_SetRenderTargetInternal(renderer, NULL);
-        SDL_RenderLogicalPresentation(renderer);
+    // !!! FIXME: fail if using a render target? Or just implicitly set the render target to NULL now?
+    if (renderer->view != &renderer->main_view) {
```

Comment:
Maybe implicitly set render target NULL? What did SDL2 do?

Reply:
SDL2 would flush the queue (possibly to a render target) and then tell the backend to present whatever might have previously be drawn to the window.

But SDL2 didn't manually draw letterboxing; in SDL2, it would be whatever the clear color was, or perhaps artifacts from the Steam Overlay, etc, if the app didn't clear it, but SDL2 didn't draw more stuff during the Present call.

To complicate this a little further: if you have a render target bound at the time on macOS, RenderPresent may do nothing. because `-[NSOpenGLContext flushBuffer]` is looking at the wrong thing.

I'm leaning towards making this an error in SDL3, since it's likely uncommon and it's probably better if we don't implicitly muck around with a bunch of state. But we _could_ notice the bound render target, flush it, unbind it, do the Present work and then bind it again before returning.

Reply:
People don't check for present errors, so if we can make this seamless, that would be good.

Reply:
Conceptually, I think it makes sense that presenting a render target is a no-op.

Reply:
This isn't terrible to just move it out of the way:

```diff
diff --git a/src/render/SDL_render.c b/src/render/SDL_render.c
index fe5cfa6c5..c5564bfa3 100644
--- a/src/render/SDL_render.c
+++ b/src/render/SDL_render.c
@@ -4964,9 +4964,9 @@ bool SDL_RenderPresent(SDL_Renderer *renderer)
 
     CHECK_RENDERER_MAGIC(renderer, false);
 
-    // !!! FIXME: fail if using a render target? Or just implicitly set the render target to NULL now?
-    if (renderer->view != &renderer->main_view) {
-        return SDL_SetError("Cannot present a render target");
+    SDL_Texture *target = renderer->target;
+    if (target) {
+        SDL_SetRenderTarget(renderer, NULL);
     }
 
     SDL_RenderLogicalPresentation(renderer);
@@ -4987,6 +4987,10 @@ bool SDL_RenderPresent(SDL_Renderer *renderer)
         presented = false;
     }
 
+    if (target) {
+        SDL_SetRenderTarget(renderer, target);
+    }
+
     if (renderer->simulate_vsync ||
         (!presented && renderer->wanted_vsync)) {
         SDL_SimulateRenderVSync(renderer);
```

...and it requires no changes to sdl2-compat, since it effectively matches SDL2 behavior.

I guess in theory if SetRenderTarget fails at this very moment (backend has lost the target or ran out of memory) this could go wrong, but in theory they're going to find out next time they try to change render targets at the app level anyhow.

Reply:
Looks good to me!

Reply:
Fixed in a78993382d83842597ea5d1676e7b9a26ec0760e.